pkg/vex/vex.go

```diff
@@ -36,62 +35,35 @@ type Statement struct {
 }
 
 type OpenVEX struct {
-	statements []Statement
-	logger     *zap.SugaredLogger
+	vex    openvex.VEX
+	logger *zap.SugaredLogger
 }
 
-func newOpenVEX(cycloneDX *ftypes.CycloneDX, vex openvex.VEX) VEX {
+func newOpenVEX(vex openvex.VEX) VEX {
 	logger := log.Logger.With(zap.String("VEX format", "OpenVEX"))
 
-	openvex.SortStatements(vex.Statements, lo.FromPtr(vex.Timestamp))
-
-	// Convert openvex.Statement to Statement
-	stmts := lo.Map(vex.Statements, func(stmt openvex.Statement, index int) Statement {
-		return Statement{
-			// TODO: add subcomponents, etc.
-			VulnerabilityID: stmt.Vulnerability,
-			Affects:         stmt.Products,
-			Status:          Status(stmt.Status),
-			Justification:   string(stmt.Justification),
-		}
-	})
-	// Reverse sorted statements so that the latest statement can come first.
-	stmts = lo.Reverse(stmts)
-
-	// If the SBOM format referenced by OpenVEX is CycloneDX
-	if cycloneDX != nil {
-		return &CycloneDX{
-			sbom:       cycloneDX,
-			statements: stmts,
-			logger:     logger,
-		}
-	}
 	return &OpenVEX{
-		statements: stmts,
-		logger:     logger,
+		vex:    vex,
+		logger: logger,
 	}
 }
 
 func (v *OpenVEX) Filter(vulns []types.DetectedVulnerability) []types.DetectedVulnerability {
 	return lo.Filter(vulns, func(vuln types.DetectedVulnerability, _ int) bool {
-		stmt, ok := lo.Find(v.statements, func(item Statement) bool {
-			return item.VulnerabilityID == vuln.VulnerabilityID
-		})
-		if !ok {
+		stmts := v.vex.Matches(vuln.VulnerabilityID, vuln.PkgRef, nil)
+		if len(stmts) == 0 {
 			return true
 		}
-		return v.affected(vuln, stmt)
-	})
-}
 
-func (v *OpenVEX) affected(vuln types.DetectedVulnerability, stmt Statement) bool {
-	if slices.Contains(stmt.Affects, vuln.PkgRef) &&
-		(stmt.Status == StatusNotAffected || stmt.Status == StatusFixed) {
-		v.logger.Infow("Filtered out the detected vulnerability", zap.String("vulnerability-id", vuln.VulnerabilityID),
-			zap.String("status", string(stmt.Status)), zap.String("justification", stmt.Justification))
-		return false
-	}
-	return true
+		// Take the effective statement
+		stmt := stmts[len(stmts)-1]
```

Comment:
Can you specify why we need get only last matched element?

If i understand correctly if we swap `statements` in [example](https://github.com/aquasecurity/trivy/blob/882acedfe487b70e4e724df56f7a006251e88c9b/pkg/vex/testdata/openvex-multiple.json#L7-L28) - we will not filter `CVE-2021-44228`.
But i am not sure that we need filter this CVE in this case.

Reply:
Added https://github.com/aquasecurity/trivy/pull/5061/commits/abb7295d398bdb8173d12f85148cb2f68805621f